includes/wc-payment-api/class-wc-payments-api-client.php

```diff
@@ -2088,20 +2089,34 @@ protected function request( $params, $api, $method, $is_site_specific = true, $u
 			);
 		}
 
-		$response = $this->http_client->remote_request(
-			[
-				'url'             => $url,
-				'method'          => $method,
-				'headers'         => apply_filters( 'wcpay_api_request_headers', $headers ),
-				'timeout'         => self::API_TIMEOUT_SECONDS,
-				'connect_timeout' => self::API_TIMEOUT_SECONDS,
-			],
-			$body,
-			$is_site_specific,
-			$use_user_token
-		);
+		$headers        = apply_filters( 'wcpay_api_request_headers', $headers );
+		$stop_trying_at = time() + self::API_TIMEOUT_SECONDS;
+		$retries        = 0;
+		$retries_limit  = array_key_exists( 'Idempotency-Key', $headers ) ? self::API_RETRIES_LIMIT : 0;
 
-		$response = apply_filters( 'wcpay_api_request_response', $response, $method, $url, $api );
+		while ( true ) {
+			$response = $this->http_client->remote_request(
+				[
+					'url'             => $url,
+					'method'          => $method,
+					'headers'         => $headers,
+					'timeout'         => self::API_TIMEOUT_SECONDS,
+					'connect_timeout' => self::API_TIMEOUT_SECONDS,
+				],
+				$body,
+				$is_site_specific,
+				$use_user_token
+			);
+
+			$response      = apply_filters( 'wcpay_api_request_response', $response, $method, $url, $api );
+			$response_code = wp_remote_retrieve_response_code( $response );
+
+			if ( $response_code || time() >= $stop_trying_at || $retries_limit === $retries ) {
```

Comment:
This approach is not refactoring-friendly, better to move `time() >= $stop_trying_at || $retries_limit === $retries` into while conditioning (with necessary tweaks) - fewer risks of infinity loops after inaccurate changes of the body.

Reply:
It's covered by unit tests, so infinite loop shouldn't slip through. Splitting condition is making it worse in my opinion, as there will be two places at which loop can end. `while ( true )` is a good indicator that end condition is _inside the loop_.

Reply:
Fair enough =) Marking as resolved!